test/Microsoft.Azure.WebJobs.Extensions.ServiceBus.Tests/Listeners/ServiceBusScaleMonitorTests.cs

```diff
@@ -0,0 +1,459 @@
+using Microsoft.Azure.WebJobs.ServiceBus.Listeners;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.Options;
+using Moq;
+using Xunit;
+using static Microsoft.Azure.ServiceBus.Message;
+
+namespace Microsoft.Azure.WebJobs.ServiceBus.UnitTests.Listeners
+{
+    public class ServiceBusScaleMonitorTests
+    {
+        //private readonly MessagingFactory _messagingFactory;
+        private readonly ServiceBusListener _listener;
+        private readonly ServiceBusScaleMonitor _scaleMonitor;
+        private readonly ServiceBusOptions _serviceBusOptions;
+        private readonly Mock<ServiceBusAccount> _mockServiceBusAccount;
+        private readonly Mock<ITriggeredFunctionExecutor> _mockExecutor;
+        private readonly Mock<MessagingProvider> _mockMessagingProvider;
+        private readonly Mock<MessageProcessor> _mockMessageProcessor;
+        private readonly ServiceBusTriggerExecutor _triggerExecutor;
+        private readonly TestLoggerProvider _loggerProvider;
+        private readonly LoggerFactory _loggerFactory;
+        private readonly string _functionId = "test-functionid";
+        private readonly string _entityPath = "test-entity-path";
+        private readonly string _testConnection = "Endpoint=sb://test.servicebus.windows.net/;SharedAccessKeyName=RootManageSharedAccessKey;SharedAccessKey=abc123=";
+
+        public ServiceBusScaleMonitorTests()
+        {
+            _mockExecutor = new Mock<ITriggeredFunctionExecutor>(MockBehavior.Strict);
+
+            MessageHandlerOptions messageOptions = new MessageHandlerOptions(ExceptionReceivedHandler);
+            MessageReceiver messageReceiver = new MessageReceiver(_testConnection, _entityPath);
+            _mockMessageProcessor = new Mock<MessageProcessor>(MockBehavior.Strict, messageReceiver, messageOptions);
+
+            _serviceBusOptions = new ServiceBusOptions
+            {
+                MessageHandlerOptions = messageOptions
+            };
+            _mockMessagingProvider = new Mock<MessagingProvider>(MockBehavior.Strict, new OptionsWrapper<ServiceBusOptions>(_serviceBusOptions));
+
+            _mockMessagingProvider
+                .Setup(p => p.CreateMessageProcessor(_entityPath, _testConnection))
+                .Returns(_mockMessageProcessor.Object);
+
+            _triggerExecutor = new ServiceBusTriggerExecutor(_mockExecutor.Object);
+            _mockServiceBusAccount = new Mock<ServiceBusAccount>(MockBehavior.Strict);
+            _mockServiceBusAccount.Setup(a => a.ConnectionString).Returns(_testConnection);
+
+            _loggerFactory = new LoggerFactory();
+            _loggerProvider = new TestLoggerProvider();
+            _loggerFactory.AddProvider(_loggerProvider);
+
+            _listener = new ServiceBusListener(_functionId, EntityType.Queue, _entityPath, false, _triggerExecutor, _serviceBusOptions, _mockServiceBusAccount.Object,
```

Comment:
these tests shouldn't need to create a listener just to create the monitor under test. Just create the monitor, and pass the mocks into it

Reply:
So the issue here is that since the `MessageReceiver` is Lazy, by default it only calls `CreateMessageReceiver` once and caches the exception. So in this test it'll throw a `MessagingEntityNotFound` exception, cache it, and continue to throw that exception for the rest of the test. This is the workaround I came up with. The alternative is I could make it a `Lazy<MessageReceiver>` with `LazyThreadSafetyMode.PublicationOnly` so it doesn't cache exceptions. It doesn't look like this would cause any issues based on [these docs](https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazythreadsafetymode?view=netframework-4.8) (assuming MessageReceiver is thread safe). Do you have any experience with this/an opinion?